### PR TITLE
Layer to show pull-backs on search map

### PR DESCRIPTION
**Pull request by lemald**
Asana ticket: [⚙️ Better Maps | Show pull-backs](https://app.asana.com/0/1200180014510248/1204946801340572/f)

This PR covers everything except for changes to the VPC for pull-backs to show their destination garage. I'm going to put that into a separate PR and release so that I don't have to start out with the new field being `optional` in Superstruct and then later change it to only `nullable`.

It's a minor thing but I don't think I've been entirely consistent with how I've written variables that involve "pull back" in their names. With snake_case I find that `pull_back` looks natural whereas for whatever reason I find `pullback` without capitalizing the "B" looks better in camelCase, but feel free to push back on that.

### CODE REVIEW

File: assets/src/components/map/controls/layersControl.tsx [hunk 1]
```diff
@@ -45,42 +52,146 @@ export const LayersControl = ({
       >
         <MapLayersIcon />
       </button>
+
+      {/* For now there's only one vehicle layer, in the future this might need to be an actual count */}
```
Comment (KaylaBrady) [inline assets/src/components/map/controls/layersControl.tsx:56]: 👍

File: assets/src/components/mapPage/mapDisplay.tsx [hunk 2]
```diff
@@ -397,14 +406,14 @@ const SelectedLocationDataLayer = ({
   )
 }
 
-const DataLayers = ({
+const SelectionLayers = ({
   selectedEntity,
-  setSelection,
+  selectVehicle,
```
Comment (KaylaBrady) [inline assets/src/components/mapPage/mapDisplay.tsx:411]: praise: I appreciate this refactor to use `setVehicle`!

File: assets/tests/components/map/controls/layersControl.test.tsx [hunk 3]
```diff
@@ -63,4 +74,97 @@ describe("LayersControl", () => {
...
+      }
+    )
+    await userEvent.click(layersControlButton.get())
+    pullbacksSwitch.get().focus()
+    await userEvent.keyboard("{Enter}")
+
+    expect(mockToggle).toHaveBeenCalledTimes(1)
+  })
+
+  test("when pull-back layer is enabled, shows pill", async () => {
```
Comment (KaylaBrady) [inline assets/tests/components/map/controls/layersControl.test.tsx:153]: suggestion (non-blocking): Maybe another test that the pill is hidden when pull-back layer is not enabled?

File: assets/tests/components/mapPage/mapDisplay.test.tsx [hunk 4]
```diff
@@ -611,6 +630,115 @@ describe("<MapDisplay />", () => {
...
+        await userEvent.click(
+          screen.getByRole("switch", { name: "Show pull-backs" })
+        )
+
+        expect(screen.getAllByRole("button", { name: "PULL-B" })).toHaveLength(
+          1
+        )
+      })
+
+      test("if a pulling back vehicle is selected, doesn't render it again in pull-backs layer", async () => {
```
Comment (KaylaBrady) [inline assets/tests/components/mapPage/mapDisplay.test.tsx:665]: praise: nice edge case consideration! I hadn't thought about this.

File: test/realtime/vehicle_test.exs [hunk 5]
```diff
@@ -579,14 +581,21 @@ defmodule Realtime.VehicleTest do
       assert Vehicle.end_of_trip_type(block, last_trip_of_block, "run2", "middle") == :pull_back
     end
 
+    test "when finished with the last trip of the block but still logged in, returns :pull_back",
```
Comment (KaylaBrady) [inline test/realtime/vehicle_test.exs:584]: question: This test name specifies "still logged in", but the change to `end_of_trip_type` doesn't seem to consider whether or not a vehicle is logged in. Should it?

**Review by KaylaBrady — COMMENTED** [review]
Looks great! Only blocking comment is question about the logic changes to `end_of_trip_type`.

File: test/realtime/vehicle_test.exs (hunk 5 above)
Comment (lemald) [reply test/realtime/vehicle_test.exs:584]: It sort of implicitly does because `end_of_trip_type/4` defaults to `:another_trip` when `run_id` is null, so the fact that this test sets a non-null run ID means that the vehicle must be logged in.

File: assets/css/map/controls/_layers_control.scss [hunk 6]
```diff
@@ -21,19 +21,53 @@
...
+  width: 22px;
+  height: 13px;
+  border-radius: 16px;
+
+  background-color: $color-strawberry-500;
+
+  text-align: center;
+  color: $white;
+  font-size: 11px;
+  line-height: 11px;
```
Comment (firestack) [inline assets/css/map/controls/_layers_control.scss:37]: suggestion: The line height is the same as the font size, so we should use unitless values

```suggestion
  line-height: 1;
```

> The used value is this unitless `<number>` multiplied by the element's own font size. The computed value is the same as the specified `<number>`. In most cases, this is the ___preferred___ way to set line-height and avoid unexpected results due to inheritance.
>
> -- https://developer.mozilla.org/en-US/docs/Web/CSS/line-height#number

File: assets/css/map/controls/_layers_control.scss [hunk 7]
```diff
@@ -21,19 +21,53 @@
...
+  left: -6px;
+  width: 22px;
+  height: 13px;
+  border-radius: 16px;
+
+  background-color: $color-strawberry-500;
+
+  text-align: center;
+  color: $white;
+  font-size: 11px;
```
[the remark this answers is not in the document]
Comment (firestack) [reply assets/css/map/controls/_layers_control.scss:36]: I'd be fine with pushing this off for a while, if we can get a ticket for potentially reworking this (and other places?) in Skate. I think this is something our design system should take care of for us honestly, so something to keep in mind for that discussion?

File: test/realtime/server_test.exs [hunk 8]
```diff
@@ -568,17 +605,21 @@ defmodule Realtime.ServerTest do
     end
 
     test "returns empty data when the route is not found", %{ets: ets} do
-      assert Server.lookup({ets, {:route_id, "2"}}) == []
+      assert Server.lookup({ets, {:route_id, "3"}}) == []
```
Comment (firestack) [inline test/realtime/server_test.exs:608]: suggestion(non-blocking): can we use -1 or something here?
```suggestion
      assert Server.lookup({ets, {:route_id, "-1"}}) == []
```

File: assets/css/map/controls/_layers_control.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (lemald) [reply assets/css/map/controls/_layers_control.scss:36]: Good idea, here's a ticket: https://app.asana.com/0/1148853526253420/1205411476253505/f

File: test/realtime/server_test.exs (hunk 8 above)
Comment (lemald) [reply test/realtime/server_test.exs:608]: Do you just mean that the `"-1"` would better denote "non-existent route ID" to someone reading the code?

Comment (firestack) [reply test/realtime/server_test.exs:608]: Yes, and I think the change here is diff and review overhead that can be avoided. The only reason this was changed was because something else took 2's spot, so ideally this wouldn't need to be changed unless something pertaining to the specific test and the implementation changed.

Comment (firestack) [reply test/realtime/server_test.exs:608]: Actually maybe it would be better to be something like this?
```suggestion
      assert Server.lookup({ets, {:route_id, "non-existent-1 "}}) == []
```

File: assets/tests/hooks/usePullbackVehicles.test.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+    // Needs to be kicked to do the effects again after the socket initializes
+    rerender()
+
+    expect(mockSocket.channel).toHaveBeenCalledTimes(1)
+    expect(mockSocket.channel).toHaveBeenCalledWith("vehicles:pull_backs:all")
+    expect(mockChannel.join).toHaveBeenCalledTimes(1)
+  })
+
+  test("does not subscribe to the pull-backs channel when the subscribe argument is flase", () => {
```
Comment (firestack) [inline assets/tests/hooks/usePullbackVehicles.test.ts:36]:
```suggestion
  test("does not subscribe to the pull-backs channel when the subscribe argument is false", () => {
```

**Review by firestack — APPROVED** [review]
